lib/Interpreter/BackendPasses.cpp

```diff
@@ -69,6 +70,52 @@ namespace {
 
 char KeepLocalGVPass::ID = 0;
 
+namespace {
+
+  // Add a suffix to the CUDA module ctor/dtor to generate an unique name.
+  // It's necessary for lazy compilation. Without suffix, cling can't
+  // differentiate the ctor/dtor of the first module from the ctor/dtor of the
+  // following modules.
+  class UnqiueCUDACtorDtorName : public ModulePass {
+    static char ID;
+
+    bool runOnGlobal(GlobalValue& GV, StringRef & ModuleName){
+      if(GV.hasName() && (GV.getName() == "__cuda_module_ctor"
+          || GV.getName() == "__cuda_module_dtor") ){
+        llvm::SmallString<128> NewFunctionName;
+        NewFunctionName.append(GV.getName());
+        NewFunctionName.append("_");
+        NewFunctionName.append(ModuleName);
+
+        for (size_t i = 0; i < NewFunctionName.size(); ++i) {
+          // Replace everything that's not [a-zA-Z0-9._] with a _. This set happens
+          // to be the set of C preprocessing numbers.
+          if (!isPreprocessingNumberBody(NewFunctionName[i]))
+            NewFunctionName[i] = '_';
+        }
+
+        GV.setName(NewFunctionName);
+
+        return true;
+      }
+
+      return false;
+    }
+
+  public:
+    UnqiueCUDACtorDtorName() : ModulePass(ID) {}
+
+    bool runOnModule(Module &M) override {
```

Comment:
I think Module should be able to be passed as const ref

Reply:
It's an override; I bet the base only has an overload taking a non-const `Module&`?

Reply:
uh, that's true here.